Mathlib/RingTheory/AlgebraicIndependent/AlgebraicClosure.lean

```diff
@@ -0,0 +1,75 @@
+
+/-!
+# Algebraic independence persists to the algebraic closure
+
+## Main results
+
+* `AlgebraicIndependent.extendScalars`: if A/S/R is a tower of algebras with S/R algebraic,
+then a family of elements in A that are algebraically independent over R remains algebraically
+independent over S, provided that S has no zero divisors.
+
+* `AlgebraicIndependent.algebraicClosure`: an algebraically independent family remains
+algebraically independent over the algebraic closure.
+-/
+
+open Function Algebra
+
+namespace AlgebraicIndependent
+
+variable {ι R S A : Type*} {x : ι → A}
+variable [CommRing R] [CommRing S] [CommRing A]
+variable [Algebra R S] [Algebra R A] [Algebra S A] [IsScalarTower R S A]
+variable [NoZeroDivisors S] (hx : AlgebraicIndependent R x)
+include hx
+
+theorem extendScalars [alg : Algebra.IsAlgebraic R S]
+    (inj : Injective (algebraMap S A)) : AlgebraicIndependent S x := by
+  nontriviality S
+  have := inj.nontrivial
+  refine algebraicIndependent_of_finite_type' inj fun t fin ind i hi ↦ ?_
+  let Rt := adjoin R (x '' t)
+  let St := adjoin S (x '' t)
+  let _ : Algebra Rt St :=
+    (Rt.inclusion (T := St.restrictScalars R) <| adjoin_le <| by exact subset_adjoin).toAlgebra
+  have : IsScalarTower Rt St A := .of_algebraMap_eq fun ⟨y, _⟩ ↦ show y = y from rfl
+  have : NoZeroDivisors St := (Set.image_eq_range _ _ ▸ ind.aevalEquiv)
+    |>.symm.injective.noZeroDivisors _ (map_zero _) (map_mul _)
+  have : NoZeroDivisors Rt := (Subalgebra.inclusion_injective _).noZeroDivisors
+    (algebraMap Rt St) (map_zero _) (map_mul _)
+  have : Algebra.IsAlgebraic Rt St := ⟨fun ⟨y, hy⟩ ↦ by
+    rw [← isAlgebraic_algHom_iff (IsScalarTower.toAlgHom Rt St A) Subtype.val_injective]
+    show IsAlgebraic Rt y
+    exact adjoin_induction (fun _ h ↦ isAlgebraic_algebraMap (⟨_, subset_adjoin h⟩ : Rt))
+      (fun z ↦ ((alg.1 z).algHom (IsScalarTower.toAlgHom R S A)).extendScalars fun _ _ eq ↦ by
+        exact hx.algebraMap_injective congr($eq.1)) (fun _ _ _ _ ↦ .add) (fun _ _ _ _ ↦ .mul) hy⟩
+  show Transcendental St (x i)
+  exact (hx.transcendental_adjoin hi).extendScalars Subtype.val_injective
+
+theorem extendScalars_of_isIntegral [Algebra.IsIntegral R S]
+    (inj : Injective (algebraMap S A)) : AlgebraicIndependent S x := by
+  nontriviality S
+  have := Module.nontrivial R S
+  exact hx.extendScalars inj
+
+theorem subalgebraAlgebraicClosure [IsDomain R] [NoZeroDivisors A] :
+    AlgebraicIndependent (Subalgebra.algebraicClosure R A) x :=
+  hx.extendScalars Subtype.val_injective
+
+protected theorem integralClosure [NoZeroDivisors A] :
+    AlgebraicIndependent (integralClosure R A) x :=
+  hx.extendScalars_of_isIntegral Subtype.val_injective
+
+omit hx in
+protected theorem algebraicClosure {F E : Type*} [Field F] [Field E] [Algebra F E] {x : ι → E}
+    (hx : AlgebraicIndependent F x) : AlgebraicIndependent (algebraicClosure F E) x :=
+  hx.extendScalars Subtype.val_injective
```

Comment:
Isn't it more general to assume `E/F` is algebraic, and then extend to `E`?

Reply:
I'm not sure what you mean? If E/F is algebraic then x can't be alg. indep. unless ι is empty, which is very restrictive / vacuous rather than general.

Reply:
Whoops, you are right. What I should have said is that we can have a tower of fields. And so we can replace `algebraicClosure F E` by any `L` such that `F/L/E` is a tower of fields, and `L/E` is algebraic.
Does that make more sense?

Reply:
But that needs an injectivity assumption.
Wouldn't the lemma that I describe be a strict generalization of this current one?

Reply:
> But that needs an injectivity assumption. Wouldn't the lemma that I describe be a strict generalization of this current one?

OK I think a `AlgebraicIndependent.extendScalars_field` may be added, which assumes in A/S/R that `S` is a field.

Reply:
It's enough to assume IsSimpleRing, since RingHom.injective has been generalized to simple rings. I've added that version.

Reply:
So now this lemma is the special case `S := algebraicClosure F E`, right?

Reply:
> It's enough to assume IsSimpleRing, since RingHom.injective has been generalized to simple rings. I've added that version.

But a commutative simple ring is automatically a field, no?

Reply:
Indeed; but probably not worth adding a variable and associated typeclasses for the Field ...

Reply:
> So now this lemma is the special case `S := algebraicClosure F E`, right?

Yes it works. [845521e](https://github.com/leanprover-community/mathlib4/pull/19708/commits/845521ed4a0cc575ef3f2c79aab317c92b9e6dd1)